Content.Server/NameIdentifier/NameIdentifierSystem.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System.Linq;
+using Content.Shared.GameTicking;
+using Content.Shared.NameIdentifier;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+
+namespace Content.Server.NameIdentifier;
+
+/// <summary>
+///     Handles unique name identifiers for entities e.g. `monkey (MK-912)`
+/// </summary>
+public class NameIdentifierSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+
+    [ViewVariables]
+    public Dictionary<NameIdentifierGroupPrototype, HashSet<int>> CurrentIds = new();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NameIdentifierComponent, ComponentInit>(OnComponentInit);
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(CleanupIds);
+
+        InitialSetupPrototypes();
+        _prototypeManager.PrototypesReloaded += OnReloadPrototypes;
+    }
+
+    public override void Shutdown()
+    {
+        base.Shutdown();
+
+        _prototypeManager.PrototypesReloaded -= OnReloadPrototypes;
+    }
+
+    /// <summary>
+    ///     Generates a new unique name/suffix for a given entity and adds it to <see cref="CurrentIds"/>
+    ///     but does not set the entity's name.
+    /// </summary>
+    public string GenerateUniqueName(EntityUid uid, NameIdentifierGroupPrototype proto)
+    {
+        var entityName = Name(uid);
+        if (!CurrentIds.TryGetValue(proto, out var set))
+            return entityName;
+
+        if (set.Count == (proto.MaxValue - proto.MinValue) + 1)
+        {
+            // Oh jeez. We're outta numbers.
+            return entityName;
+        }
+
+        // This is kind of inefficient with very large amounts of entities but its better than any other method
+        // I could come up with.
+
+        var randomVal = _robustRandom.Next(proto.MinValue, proto.MaxValue);
+        while (set.Contains(randomVal))
+        {
+            randomVal = _robustRandom.Next(proto.MinValue, proto.MaxValue);
+        }
```

Comment:
Could fallback to just iterate from smallest to highest, alternatively could use the random value as the start index and iterate from there, then wraparound at highest?

Reply:
when would it fallback? after like N failed attempts or something?

Reply:
will leave it as is, imo